src/cli/subcommands/snapshot_cmd.rs

```diff
@@ -162,32 +170,70 @@ async fn validate(
         &store,
     )
     .await?;
-
+    let chain_data_root = TempDir::new()?;
     let chain_store = Arc::new(ChainStore::new(
         store,
         config.chain.clone(),
         &genesis,
-        TempDir::new()?.path(),
+        chain_data_root.path(),
     )?);
 
     let ts = chain_store.tipset_from_keys(&TipsetKeys::new(chain_store.db.roots()))?;
 
     validate_links_and_genesis_traversal(
         &chain_store,
-        ts,
+        &ts,
         chain_store.blockstore(),
         *recent_stateroots,
         &Tipset::from(genesis),
         &config.chain.network,
     )
     .await?;
 
+    if let Some(validate_from) = *validate_tipsets {
+        // Set proof parameter data dir
+        if cns::FETCH_PARAMS {
+            crate::utils::proofs_api::paramfetch::set_proofs_parameter_cache_dir_env(
+                &config.client.data_dir,
+            );
+        }
+        // Initialize StateManager
+        let state_manager = Arc::new(StateManager::new(
+            chain_store,
+            Arc::clone(&config.chain),
+            Arc::new(crate::interpreter::RewardActorMessageCalc),
+        )?);
+        ensure_params_downloaded().await?;
+        // Prepare tipset stream to validate
+        let heaviest_epoch = ts.epoch();
+        let end_tipset = state_manager
+            .chain_store()
+            .tipset_by_height(heaviest_epoch, ts, false)
+            .context(format!("couldn't get a tipset at height {heaviest_epoch}"))?;
+        let tipsets = itertools::unfold(Some(end_tipset), |tipset| {
+            let child = tipset.take()?;
+            *tipset = state_manager
+                .chain_store()
+                .tipset_from_keys(child.parents())
+                .ok();
+            Some(child)
```

Comment:
this is copy-pasted - can you refactor to a common `get_all_ancestors` function please